src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -196,13 +197,56 @@ protected Solution SetCurrentSolution(Solution solution)
             while (true)
             {
                 var newSolution = solution.WithNewWorkspace(this, currentSolution.WorkspaceVersion + 1);
-                var replacedSolution = Interlocked.CompareExchange(ref _latestSolution, newSolution, currentSolution);
-                if (replacedSolution == currentSolution)
+                var oldSolution = Interlocked.CompareExchange(ref _latestSolution, newSolution, currentSolution);
+                if (oldSolution == currentSolution)
                 {
                     return newSolution;
                 }
 
-                currentSolution = replacedSolution;
+                currentSolution = oldSolution;
+            }
+        }
+
+        /// <summary>
+        /// Applies specified transformation to <see cref="CurrentSolution"/>, updates <see cref="CurrentSolution"/> to the new value and raises a workspace change event of the specified kind.
+        /// </summary>
+        /// <param name="transformation">Solution transformation.</param>
+        /// <param name="kind">The kind of workspace change event to raise.</param>
+        /// <param name="projectId">The id of the project updated by <paramref name="transformation"/> to be passed to the workspace change event.</param>
+        /// <param name="documentId">The id of the document updated by <paramref name="transformation"/> to be passed to the workspace change event.</param>
+        /// <returns>True if <see cref="CurrentSolution"/> was set to the transformed solution, false if the transformation did not change the solution.</returns>
+        internal bool SetCurrentSolution(Func<Solution, Solution> transformation, WorkspaceChangeKind kind, ProjectId? projectId = null, DocumentId? documentId = null)
+        {
+            Contract.ThrowIfNull(transformation);
+
+            var currentSolution = Volatile.Read(ref _latestSolution);
+
+            while (true)
+            {
+                var transformedSolution = transformation(currentSolution);
+                if (transformedSolution == currentSolution)
+                {
+                    return false;
+                }
+
+                var newSolution = transformedSolution.WithNewWorkspace(this, currentSolution.WorkspaceVersion + 1);
+
+                Solution oldSolution;
+                using (_serializationLock.DisposableWait())
+                {
+                    oldSolution = Interlocked.CompareExchange(ref _latestSolution, newSolution, currentSolution);
```

Comment:
It feels a bit strange that we're doing all this lock-free stuff, and in the end we're still taking a lock around it anyways. It's not wrong, just feels strange.